rfcs/2021-03-26-6517-end-to-end-acknowledgement.md

```diff
@@ -0,0 +1,344 @@
+}
+
+struct EventFinalizer {
+    status: EventStatus,
+    sources: Box<[Arc<BatchNotifier>]>,
+}
+
+struct BatchNotifier {
+    status: Mutex<BatchStatus>,
+    notifier: tokio::sync::oneshot::Sender<BatchStatus>,
+    identifier: ImmStr,
+}
+
+struct EventFinalization {
+    id: EventId,
+    status: EventStatus,
+}
+
+enum BatchStatus {
+    Delivered,
+    Failed,
+}
+
+enum EventStatus {
+    Dropped, // default status
+    Delivered,
+    Failed,
+    NoOp,
+}
+```
+
+### Source configuration
+
+When building a source, the topology system will provide it with an
+unique source identifier. The source is responsible for creating its own
+MPSC channel pair to provide to events. This allows sources to create a
+separate channel for each sender or batch, as convenient for the source
+protocol, and then `await` on the receiver to collect the finalization
+events. Since the size of each batch is known in advance, this may be a
+bounded channel with the length fixed to the size of the source batch.
+When the last event in the batch is dropped, the sender will be closed
+and the receiver will be signalled that all events are completed.
+
+Since this is making the parameter list for `trait SourceConfig::build`
+increasingly unweidly, a new context structure will be introduced to
+carry the data.
+
+```rust
+struct SourceContext<'a> {
+    name: &'a str,
+    identifier: Box<str>,
+    globals: &'a GlobalOptions,
+    shutdown: ShutdownSignal,
+    out: Pipeline,
+}
+
+trait SourceConfig {
+    async fn build(&self, context: SourceContext<'_>) -> crate::Result<sources::Source>;
+}
+```
+
+### Sink configuration
+
+A new global configuration setting will be added to all sources,
+flagging one or more sinks as "authoritative". This setting will change
+the behavior of finalization as described above. When an event is
+delivered to an authoritative sink, its finalization status is
+immediately delivered to all sources of the event. In order to prevent
+another status from being delivered at a later point, the status of the
+event is changed to "no-op", the same as if it was merged into another
+event.
+
+```rust
+struct SinkOuter {
+    // … existing fields …
+
+    #[serde(default)]
+    authoritative: bool,
+}
+```
+
+## Doc-level Proposal
+
+The following additional source configuration will be added:
+
+```toml
+# Global enable option
+acknowledgements = true
+
+[sources.my_source_id]
+  # Per-source enable option
+  acknowledgements = true
+```
+
+We would also need to document when acknowledgement happens for each source.
+
+## Rationale
+
+The above structure provides for several considerations:
+
+1.  This the minimum amount of data that can be added to the metadata to
+    fully support this feature, amounting to a single shared reference
+    as the `Option` is optimized into the `Arc`.
+
+2.  The use of `Arc` reference counting for finalization prevents events
+    from "escaping" without providing a status indication.
```

Comment:
I would support putting finalization logic in `Drop`. I _think_ some synchronization could be elided in that case, with the downside being we don't necessarily control when `Drop` is executed.